### PR TITLE
test: Retrieve the private interface in an Eventually

### LINKED ISSUE #16479: CI: K8sDatapathConfig Transparent encryption DirectRouting Check connectivity with transparent encryption and direct routing: Unable to determine private iface
Hit in an unrelated pull request:
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.19-kernel-5.4/299/testReport/junit/Suite-k8s-1/19/K8sDatapathConfig_Transparent_encryption_DirectRouting_Check_connectivity_with_transparent_encryption_and_direct_routing/
[40d7286d_K8sDatapathConfig_Transparent_encryption_DirectRouting_Check_connectivity_with_transparent_encryption_and_direct_routing.zip](https://github.com/cilium/cilium/files/6618841/40d7286d_K8sDatapathConfig_Transparent_encryption_DirectRouting_Check_connectivity_with_transparent_encryption_and_direct_routing.zip)

### Stacktrace
```
/home/jenkins/workspace/Cilium-PR-K8s-1.19-kernel-5.4/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:518
Unable to determine private iface
Expected
    <*errors.errorString | 0xc00031d810>: {
        s: "Failed to retrieve iface by IP addr: Failed to exec \"ip -j a s  | jq -r '.[] | select(.addr_info[] | .local == \\\"192.168.36.11\\\") | .ifname'\" cmd on \"k8s1\" node:  (exit status 4) output: cmd: kubectl exec -n kube-system log-gatherer-9zc85 -- ip -j a s  | jq -r '.[] | select(.addr_info[] | .local == \"192.168.36.11\") | .ifname'\nExitcode: 4 \nErr: exit status 4\nStdout:\n \t \nStderr:\n \t parse error: Unfinished JSON term at EOF at line 1, column 7673\n\t \n",
    }
to be nil
/home/jenkins/workspace/Cilium-PR-K8s-1.19-kernel-5.4/src/github.com/cilium/cilium/test/k8sT/DatapathConfiguration.go:692
```
### Standard Output
```
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Cilium pods: [cilium-gmsl7 cilium-vrx98]
Netpols loaded: 
CiliumNetworkPolicies loaded: 
Endpoint Policy Enforcement:
Pod                           Ingress   Egress
grafana-d69c97b9b-j2hzh                 
prometheus-655fb888d7-sxxhs             
coredns-867bf6789f-gxt2b                
Cilium agent 'cilium-vrx98': Status: Ok  Health: Ok Nodes "" ContinerRuntime:  Kubernetes: Ok KVstore: Ok Controllers: Total 20 Failed 0
```

### PR DESCRIPTION
**Pull request by christarazi**
Upon sysdump inspection of
https://github.com/cilium/cilium/issues/16479, it's possible that the
failures in retrieving the private interface are time-sensative. As
https://github.com/cilium/cilium/issues/16479#issuecomment-885917267
points out, the output of `ip a` actually showed that the interface the
test was looking for existed, so it's very likely a timing issue. This
commit attempts to fix this flake by retrying until we are able to
extract out the interface.

Fixes: https://github.com/cilium/cilium/issues/16479

### CODE REVIEW

**Comment by christarazi** [comment]
test-1.16-netnext

**Comment by christarazi** [comment]
test-1.20-4.19

**Comment by christarazi** [comment]
test-1.19-5.4

**Comment by pchaigno** [comment]
Do we have any idea why this would be happening? When the test fails, it's usually not the first test to run, so the enp0s8 interface has been set up for a while already.

**Comment by christarazi** [comment]
@pchaigno It might have something to do with Cilium being re-installed and it re-configuring the devices. I tried reproducing locally but I couldn't even when forcing  the relevant previous test to run as https://github.com/cilium/cilium/issues/16479#issuecomment-884664149 pointed out. I can't think of any other explanation because the test actually spits out `ip a` when we fail to find the private iface and that error msg contains `enp0s8`. 

Actually another thing that I just thought about is that it could be netlink being unreliable. It's possible that the first call to `ip a` led to an empty response back hence `enp0s8` missing. An `Eventually` should cover both cases. I don't think this a Cilium bug in any way, so I don't think we are papering over anything.

**Comment by pchaigno** [comment]
> Actually another thing that I just thought about is that it could be netlink being unreliable. It's possible that the first call to `ip a` led to an empty response back hence `enp0s8` missing.

That's what I had in mind. Ideally, we'd print, in the error message, the exact `ip a` output we've used. I just couldn't figure out a clean way to do that at the time :-/

> An `Eventually` should cover both cases. I don't think this is a Cilium bug in any way, so I don't think we are papering over anything.

Yep, you're probably right.

**Review by nbusseneau — APPROVED** [review]
:shipit:

**Comment by nbusseneau** [comment]
Does this PR really need a review from @cilium/bpf?

**Comment by pchaigno** [comment]
> Does this PR really need a review from @cilium/bpf?

No, but we can't remove the team review request, so Chris removed the reviewer. I guess it can be merged now if tests are passing.

**Comment by nbusseneau** [comment]
I'll let Chris mark as `ready-to-merge` if he feels it's ready, otherwise I suggest we run `test-1.19-5.4` a few times since it's the only job I've seen fail with this so far.

**Comment by christarazi** [comment]
test-1.19-5.4

**Review by joestringer — APPROVED** [review]

**Review by nebril — APPROVED** [review]

**Comment by christarazi** [comment]
test-1.19-5.4

**Pull request merged**
